pkg/middleware/stored_session.go

```diff
@@ -111,7 +116,7 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 // is older than the refresh period.
 // Success or fail, we will then validate the session.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
-	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
+	if !session.IsExpiredOn(time.Now().Add(s.sessionExpiryOffset)) && (s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod) {
```

Comment:
This poses an interesting question, previously, if the refresh failed, we would remove the session, are we still going to remove the session if the refresh fails and the session hasn't actually expired yet?

Reply:
I think we kill the session?

If we don't kill the session, would we try to refresh again on every subsequent request until the session or access token is expired? This protects us from intermittent refresh endpoint failures (non-security, e.g. network). But comes with a severe performance & security issue if the session really is invalid.

I say we assume if refresh fails, something administratively was done on the IdP side to invalidate the token & we kill the session immediately for security.